File: backend/src/api/progress.py
"""Progress API endpoints for chapter tracking and dashboard."""

from datetime import datetime
from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from src.lib.database import get_db
from src.models.chapter import Chapter
from src.models.progress import ProgressStatus
from src.services import progress_service
from src.api.auth import get_current_user, ErrorResponse

router = APIRouter()


# Request/Response Models
class ChapterProgress(BaseModel):
    """Chapter progress response model."""

    chapter_id: str
    module_id: str
    status: str
    completed_at: datetime | None
    reading_time_seconds: int
    last_position: str | None
    updated_at: datetime

    class Config:
        from_attributes = True


class ProgressSummary(BaseModel):
    """Progress summary response model."""

    total_chapters: int
    completed_chapters: int
    in_progress_chapters: int
    total_reading_time_minutes: int
    chapters: list[ChapterProgress]


class UpdateProgressRequest(BaseModel):
    """Request model for progress update."""

    status: str | None = Field(None, pattern="^(in_progress|completed)$")
    last_position: str | None = None
    additional_reading_time: int | None = Field(None, ge=0)


class HeartbeatRequest(BaseModel):
    """Request model for reading heartbeat."""

    chapter_id: str
    position: str | None = None
    duration_seconds: int = Field(default=30, ge=1, le=300)


class OverallProgress(BaseModel):
    """Overall progress stats."""

    completed: int
    in_progress: int
    not_started: int
    total_chapters: int
    percent_complete: float


class ModuleProgress(BaseModel):
    """Module progress stats."""

    module_id: str
    module_title: str
    completed: int
    total: int
    percent_complete: float


class RecentActivity(BaseModel):
    """Recent activity item."""

    chapter_id: str
    chapter_title: str
    action: str
    timestamp: str


class Dashboard(BaseModel):
    """Dashboard response model."""

    overall_progress: OverallProgress
    module_progress: list[ModuleProgress]
    recent_activity: list[RecentActivity]
    total_reading_time_minutes: int
    current_streak: int


class NextChapter(BaseModel):
    """Next chapter recommendation."""

    chapter_id: str
    title: str
    module_id: str
    reason: str


class SkipSuggestion(BaseModel):
    """Skip suggestion item."""

    chapter_id: str
    title: str
    reason: str


class FocusArea(BaseModel):
    """Focus area recommendation."""

    module_id: str
    module_title: str
    priority: str


class Recommendations(BaseModel):
    """Recommendations response model."""

    next_chapter: NextChapter | None
    skip_suggestions: list[SkipSuggestion]
    focus_areas: list[FocusArea]


# Helper function
async def verify_chapter_exists(db: AsyncSession, chapter_id: str) -> Chapter:
    """Verify a chapter exists and return it."""
    result = await db.execute(select(Chapter).where(Chapter.id == chapter_id))
    chapter = result.scalar_one_or_none()
    if not chapter:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail={"code": "CHAPTER_NOT_FOUND", "message": f"Chapter '{chapter_id}' not found"},
        )
    return chapter


# Endpoints
@router.get(
    "",
    response_model=ProgressSummary,
    responses={
        401: {"model": ErrorResponse, "description": "Not authenticated"},
    },
)
async def get_all_progress(
    db: Annotated[AsyncSession, Depends(get_db)],
    auth: Annotated[tuple, Depends(get_current_user)],
    module_id: str | None = Query(None, description="Filter by module ID"),
) -> ProgressSummary:
    """Get all progress for current user."""
    _, user = auth

    # Get summary
    summary = await progress_service.get_progress_summary(db, user.id, module_id)

    # Get all chapters
    chapter_query = select(Chapter)
    if module_id:
        chapter_query = chapter_query.where(Chapter.module_id == module_id)
    result = await db.execute(chapter_query)
    all_chapters = list(result.scalars().all())

    # Get progress records
    progress_records = await progress_service.get_user_progress(db, user.id, module_id)
    progress_map = {p.chapter_id: p for p in progress_records}

    # Build chapter progress list
    chapters = []
    for chapter in all_chapters:
        progress = progress_map.get(chapter.id)
        if progress:
            chapters.append(ChapterProgress(
                chapter_id=chapter.id,
                module_id=chapter.module_id,
                status=progress.status,
                completed_at=progress.completed_at,
                reading_time_seconds=progress.reading_time_seconds,
                last_position=progress.last_position,
                updated_at=progress.updated_at,
            ))
        else:
            chapters.append(ChapterProgress(
                chapter_id=chapter.id,
                module_id=chapter.module_id,
                status=ProgressStatus.NOT_STARTED,
                completed_at=None,
                reading_time_seconds=0,
                last_position=None,
                updated_at=datetime.utcnow(),
            ))

    return ProgressSummary(
        total_chapters=summary["total_chapters"],
        completed_chapters=summary["completed_chapters"],
        in_progress_chapters=summary["in_progress_chapters"],
        total_reading_time_minutes=summary["total_reading_time_minutes"],
        chapters=chapters,
    )


@router.get(
    "/chapters/{chapter_id}",
    response_model=ChapterProgress,
    responses={
        401: {"model": ErrorResponse, "description": "Not authenticated"},
        404: {"model": ErrorResponse, "description": "Chapter not found"},
    },
)
async def get_chapter_progress(
    chapter_id: str,
    db: Annotated[AsyncSession, Depends(get_db)],
    auth: Annotated[tuple, Depends(get_current_user)],
) -> ChapterProgress:
    """Get progress for a specific chapter."""
    _, user = auth

    # Verify chapter exists
    chapter = await verify_chapter_exists(db, chapter_id)

    # Get progress
    progress = await progress_service.get_chapter_progress(db, user.id, chapter_id)

    if progress:
        return ChapterProgress(
            chapter_id=chapter.id,
            module_id=chapter.module_id,
            status=progress.status,
            completed_at=progress.completed_at,
            reading_time_seconds=progress.reading_time_seconds,
            last_position=progress.last_position,
            updated_at=progress.updated_at,
        )
    else:
        return ChapterProgress(
            chapter_id=chapter.id,
            module_id=chapter.module_id,
            status=ProgressStatus.NOT_STARTED,
            completed_at=None,
            reading_time_seconds=0,
            last_position=None,
            updated_at=datetime.utcnow(),
        )


@router.patch(
    "/chapters/{chapter_id}",
    response_model=ChapterProgress,
    responses={
        400: {"model": ErrorResponse, "description": "Invalid input"},
        401: {"model": ErrorResponse, "description": "Not authenticated"},
        404: {"model": ErrorResponse, "description": "Chapter not found"},
    },
)
async def update_chapter_progress(
    chapter_id: str,
    body: UpdateProgressRequest,
    db: Annotated[AsyncSession, Depends(get_db)],
    auth: Annotated[tuple, Depends(get_current_user)],
) -> ChapterProgress:
    """Update progress for a chapter."""
    _, user = auth

    # Verify chapter exists
    chapter = await verify_chapter_exists(db, chapter_id)

    # Update progress
    progress = await progress_service.update_chapter_progress(
        db=db,
        user_id=user.id,
        chapter_id=chapter_id,
        status=body.status,
        last_position=body.last_position,
        additional_reading_time=body.additional_reading_time,
    )

    return ChapterProgress(
        chapter_id=chapter.id,
        module_id=chapter.module_id,
        status=progress.status,
        completed_at=progress.completed_at,
        reading_time_seconds=progress.reading_time_seconds,
        last_position=progress.last_position,
        updated_at=progress.updated_at,
    )


@router.post(
    "/chapters/{chapter_id}/complete",
    response_model=ChapterProgress,
    responses={
        401: {"model": ErrorResponse, "description": "Not authenticated"},
        404: {"model": ErrorResponse, "description": "Chapter not found"},
    },
)
async def mark_chapter_complete(
    chapter_id: str,
    db: Annotated[AsyncSession, Depends(get_db)],
    auth: Annotated[tuple, Depends(get_current_user)],
) -> ChapterProgress:
    """Mark a chapter as complete."""
    _, user = auth

    # Verify chapter exists
    chapter = await verify_chapter_exists(db, chapter_id)

    # Mark complete
    progress = await progress_service.mark_chapter_complete(db, user.id, chapter_id)

    return ChapterProgress(
        chapter_id=chapter.id,
        module_id=chapter.module_id,
        status=progress.status,
        completed_at=progress.completed_at,
        reading_time_seconds=progress.reading_time_seconds,
        last_position=progress.last_position,
        updated_at=progress.updated_at,
    )


@router.delete(
    "/chapters/{chapter_id}/complete",
    response_model=ChapterProgress,
    responses={
        401: {"model": ErrorResponse, "description": "Not authenticated"},
    },
)
async def unmark_chapter_complete(
    chapter_id: str,
    db: Annotated[AsyncSession, Depends(get_db)],
    auth: Annotated[tuple, Depends(get_current_user)],
) -> ChapterProgress:
    """Unmark chapter completion."""
    _, user = auth

    # Get chapter (don't require it exists - might have been deleted)
    result = await db.execute(select(Chapter).where(Chapter.id == chapter_id))
    chapter = result.scalar_one_or_none()

    # Unmark complete
    progress = await progress_service.unmark_chapter_complete(db, user.id, chapter_id)

    return ChapterProgress(
        chapter_id=chapter_id,
        module_id=chapter.module_id if chapter else "",
        status=progress.status,
        completed_at=progress.completed_at,
        reading_time_seconds=progress.reading_time_seconds,
        last_position=progress.last_position,
        updated_at=progress.updated_at,
    )


@router.get(
    "/dashboard",
    response_model=Dashboard,
    responses={
        401: {"model": ErrorResponse, "description": "Not authenticated"},
    },
)
async def get_dashboard(
    db: Annotated[AsyncSession, Depends(get_db)],
    auth: Annotated[tuple, Depends(get_current_user)],
) -> Dashboard:
    """Get progress dashboard data."""
    _, user = auth

    data = await progress_service.get_dashboard_data(db, user.id)

    return Dashboard(
        overall_progress=OverallProgress(**data["overall_progress"]),
        module_progress=[ModuleProgress(**mp) for mp in data["module_progress"]],
        recent_activity=[RecentActivity(**ra) for ra in data["recent_activity"]],
        total_reading_time_minutes=data["total_reading_time_minutes"],
        current_streak=data["current_streak"],
    )


@router.get(
    "/recommendations",
    response_model=Recommendations,
    responses={
        401: {"model": ErrorResponse, "description": "Not authenticated"},
    },
)
async def get_recommendations(
    db: Annotated[AsyncSession, Depends(get_db)],
    auth: Annotated[tuple, Depends(get_current_user)],
) -> Recommendations:
    """Get personalized content recommendations."""
    _, user = auth

    data = await progress_service.get_recommendations(db, user)

    return Recommendations(
        next_chapter=NextChapter(**data["next_chapter"]) if data["next_chapter"] else None,
        skip_suggestions=[SkipSuggestion(**ss) for ss in data["skip_suggestions"]],
        focus_areas=[FocusArea(**fa) for fa in data["focus_areas"]],
    )


@router.post(
    "/heartbeat",
    responses={
        401: {"model": ErrorResponse, "description": "Not authenticated"},
    },
)
async def heartbeat(
    body: HeartbeatRequest,
    db: Annotated[AsyncSession, Depends(get_db)],
    auth: Annotated[tuple, Depends(get_current_user)],
) -> dict[str, str]:
    """Update reading activity (called periodically)."""
    _, user = auth

    await progress_service.record_heartbeat(
        db=db,
        user_id=user.id,
        chapter_id=body.chapter_id,
        position=body.position,
        duration_seconds=body.duration_seconds,
    )

    return {"status": "recorded"}

File: frontend/src/services/progress.ts
/**
 * Progress tracking service.
 *
 * DISABLED MODE: Returns mock data when backend is unavailable.
 * Progress is not persisted without a backend.
 */

import { api } from './api';

// Types
export interface ChapterProgress {
  chapterId: string;
  moduleId: string;
  status: 'not_started' | 'in_progress' | 'completed';
  completedAt: string | null;
  readingTimeSeconds: number;
  lastPosition: string | null;
  updatedAt: string;
}

export interface ProgressSummary {
  totalChapters: number;
  completedChapters: number;
  inProgressChapters: number;
  totalReadingTimeMinutes: number;
  chapters: ChapterProgress[];
}

export interface OverallProgress {
  completed: number;
  inProgress: number;
  notStarted: number;
  totalChapters: number;
  percentComplete: number;
}

export interface ModuleProgress {
  moduleId: string;
  moduleTitle: string;
  completed: number;
  total: number;
  percentComplete: number;
}

export interface RecentActivity {
  chapterId: string;
  chapterTitle: string;
  action: 'started' | 'continued' | 'completed';
  timestamp: string;
}

export interface Dashboard {
  overallProgress: OverallProgress;
  moduleProgress: ModuleProgress[];
  recentActivity: RecentActivity[];
  totalReadingTimeMinutes: number;
  currentStreak: number;
}

export interface NextChapter {
  chapterId: string;
  title: string;
  moduleId: string;
  reason: string;
}

export interface SkipSuggestion {
  chapterId: string;
  title: string;
  reason: string;
}

export interface FocusArea {
  moduleId: string;
  moduleTitle: string;
  priority: 'high' | 'medium' | 'low';
}

export interface Recommendations {
  nextChapter: NextChapter | null;
  skipSuggestions: SkipSuggestion[];
  focusAreas: FocusArea[];
}

// Mock data for when backend is unavailable
const mockDashboard: Dashboard = {
  overallProgress: {
    completed: 0,
    inProgress: 0,
    notStarted: 20,
    totalChapters: 20,
    percentComplete: 0,
  },
  moduleProgress: [
    { moduleId: 'intro', moduleTitle: 'Introduction', completed: 0, total: 3, percentComplete: 0 },
    { moduleId: 'module-1-ros2', moduleTitle: 'Module 1: ROS 2', completed: 0, total: 4, percentComplete: 0 },
    { moduleId: 'module-2-simulation', moduleTitle: 'Module 2: Simulation', completed: 0, total: 4, percentComplete: 0 },
    { moduleId: 'module-3-isaac', moduleTitle: 'Module 3: NVIDIA Isaac', completed: 0, total: 4, percentComplete: 0 },
    { moduleId: 'module-4-vla', moduleTitle: 'Module 4: VLA', completed: 0, total: 4, percentComplete: 0 },
  ],
  recentActivity: [],
  totalReadingTimeMinutes: 0,
  currentStreak: 0,
};

const mockRecommendations: Recommendations = {
  nextChapter: {
    chapterId: 'intro',
    title: 'Introduction to Physical AI',
    moduleId: 'intro',
    reason: 'Start your learning journey here!',
  },
  skipSuggestions: [],
  focusAreas: [],
};

const mockProgressSummary: ProgressSummary = {
  totalChapters: 20,
  completedChapters: 0,
  inProgressChapters: 0,
  totalReadingTimeMinutes: 0,
  chapters: [],
};

const createMockChapterProgress = (chapterId: string): ChapterProgress => ({
  chapterId,
  moduleId: 'unknown',
  status: 'not_started',
  completedAt: null,
  readingTimeSeconds: 0,
  lastPosition: null,
  updatedAt: new Date().toISOString(),
});

class ProgressService {
  private backendDisabled = false;

  /**
   * Check if backend is available and cache result.
   */
  private async checkBackend(): Promise<boolean> {
    if (this.backendDisabled) return false;

    try {
      const available = await api.isBackendAvailable();
      if (!available) {
        this.backendDisabled = true;
        console.info('Progress tracking disabled: backend not available');
      }
      return available;
    } catch {
      this.backendDisabled = true;
      return false;
    }
  }

  /**
   * Get all progress for current user.
   */
  async getAllProgress(moduleId?: string): Promise<ProgressSummary> {
    if (!(await this.checkBackend())) {
      return mockProgressSummary;
    }

    try {
      const params = moduleId ? `?module_id=${moduleId}` : '';
      return await api.get<ProgressSummary>(`/progress${params}`);
    } catch {
      return mockProgressSummary;
    }
  }

  /**
   * Get progress for a specific chapter.
   */
  async getChapterProgress(chapterId: string): Promise<ChapterProgress> {
    if (!(await this.checkBackend())) {
      return createMockChapterProgress(chapterId);
    }

    try {
      return await api.get<ChapterProgress>(`/progress/chapters/${chapterId}`);
    } catch {
      return createMockChapterProgress(chapterId);
    }
  }

  /**
   * Update chapter progress.
   */
  async updateChapterProgress(
    chapterId: string,
    data: {
      status?: 'in_progress' | 'completed';
      lastPosition?: string;
      additionalReadingTime?: number;
    }
  ): Promise<ChapterProgress> {
    if (!(await this.checkBackend())) {
      // Return mock updated progress
      return {
        ...createMockChapterProgress(chapterId),
        status: data.status || 'not_started',
        lastPosition: data.lastPosition || null,
      };
    }

    try {
      return await api.patch<ChapterProgress>(`/progress/chapters/${chapterId}`, {
        status: data.status,
        last_position: data.lastPosition,
        additional_reading_time: data.additionalReadingTime,
      });
    } catch {
      return createMockChapterProgress(chapterId);
    }
  }

  /**
   * Mark a chapter as complete.
   */
  async markChapterComplete(chapterId: string): Promise<ChapterProgress> {
    if (!(await this.checkBackend())) {
      return {
        ...createMockChapterProgress(chapterId),
        status: 'completed',
        completedAt: new Date().toISOString(),
      };
    }

    try {
      return await api.post<ChapterProgress>(`/progress/chapters/${chapterId}/complete`);
    } catch {
      return createMockChapterProgress(chapterId);
    }
  }

  /**
   * Unmark chapter completion.
   */
  async unmarkChapterComplete(chapterId: string): Promise<ChapterProgress> {
    if (!(await this.checkBackend())) {
      return createMockChapterProgress(chapterId);
    }

    try {
      return await api.delete<ChapterProgress>(`/progress/chapters/${chapterId}/complete`);
    } catch {
      return createMockChapterProgress(chapterId);
    }
  }

  /**
   * Get dashboard data.
   */
  async getDashboard(): Promise<Dashboard> {
    if (!(await this.checkBackend())) {
      return mockDashboard;
    }

    try {
      return await api.get<Dashboard>('/progress/dashboard');
    } catch {
      return mockDashboard;
    }
  }

  /**
   * Get recommendations.
   */
  async getRecommendations(): Promise<Recommendations> {
    if (!(await this.checkBackend())) {
      return mockRecommendations;
    }

    try {
      return await api.get<Recommendations>('/progress/recommendations');
    } catch {
      return mockRecommendations;
    }
  }

  /**
   * Send reading heartbeat.
   */
  async heartbeat(
    chapterId: string,
    position?: string,
    durationSeconds: number = 30
  ): Promise<void> {
    if (!(await this.checkBackend())) {
      // Silently ignore heartbeats when backend is unavailable
      return;
    }

    try {
      await api.post('/progress/heartbeat', {
        chapter_id: chapterId,
        position,
        duration_seconds: durationSeconds,
      });
    } catch {
      // Silently fail heartbeats
    }
  }
}

// Export singleton instance
export const progressService = new ProgressService();

// Export class for testing
export { ProgressService };
